File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditRevertHelpView.java
package org.wikipedia.descriptions;

import android.content.Context;
import android.net.Uri;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.text.style.BulletSpan;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.databinding.ViewDescriptionEditRevertHelpBinding;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.StringUtil;

public class DescriptionEditRevertHelpView extends ScrollView {
    public DescriptionEditRevertHelpView(@NonNull Context context, @NonNull String qNumber) {
        super(context);
        init(qNumber);
    }

    @SuppressWarnings("checkstyle:magicnumber") private void init(@NonNull String qNumber) {
        final ViewDescriptionEditRevertHelpBinding binding = ViewDescriptionEditRevertHelpBinding.bind(this);
        final TextView helpText = binding.viewDescriptionEditRevertHelpContents;

        helpText.setMovementMethod(new LinkMovementMethod());

        Spanned helpStr = StringUtil.fromHtml(getString(R.string.description_edit_revert_help_body)
                .replaceAll(":revertSubtitle", getString(R.string.description_edit_revert_subtitle))
                .replaceAll(":revertIntro", getString(R.string.description_edit_revert_intro))
                .replaceAll(":revertHistory",
                        getContext().getString(R.string.description_edit_revert_history, getHistoryUri(qNumber))));

        int gapWidth = DimenUtil.roundedDpToPx(8);
        SpannableString revertReason1 = new SpannableString(StringUtil.fromHtml(getContext().getString(R.string.description_edit_revert_reason1, getString(R.string.wikidata_description_guide_url))));
        SpannableString revertReason2 = new SpannableString(StringUtil.fromHtml(getString(R.string.description_edit_revert_reason2)));
        revertReason1.setSpan(new BulletSpan(gapWidth), 0, revertReason1.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        revertReason2.setSpan(new BulletSpan(gapWidth), 0, revertReason2.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        SpannableString helpSpan = new SpannableString(TextUtils.expandTemplate(helpStr, revertReason1, revertReason2));
        helpText.setText(helpSpan);
    }

    private Uri getHistoryUri(@NonNull String qNumber) {
        return new Uri.Builder()
                .scheme(WikipediaApp.getInstance().getWikiSite().scheme())
                .authority("m.wikidata.org")
                .appendPath("wiki")
                .appendPath("Special:History")
                .appendPath(qNumber)
                .build();
    }

    private String getString(@StringRes int id) {
        return getContext().getString(id);
    }
}

File: app/src/main/java/org/wikipedia/descriptions/DescriptionEditRevertHelpView.kt
package org.wikipedia.descriptions

import android.content.Context
import android.net.Uri
import android.text.SpannableString
import android.text.TextUtils
import android.text.method.LinkMovementMethod
import android.text.style.BulletSpan
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.ScrollView
import androidx.core.text.buildSpannedString
import androidx.core.text.inSpans
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.databinding.ViewDescriptionEditRevertHelpBinding
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.StringUtil

class DescriptionEditRevertHelpView constructor(context: Context, attrs: AttributeSet? = null) : ScrollView(context, attrs) {

    private var qNumber: String = ""

    constructor(context: Context, qNumber: String) : this(context, null) {
        this.qNumber = qNumber
    }

    init {
        val binding = ViewDescriptionEditRevertHelpBinding.inflate(LayoutInflater.from(context), this)
        binding.helpText.movementMethod = LinkMovementMethod()
        val helpStr = StringUtil.fromHtml(context.getString(R.string.description_edit_revert_help_body)
                .replace(":revertSubtitle".toRegex(), context.getString(R.string.description_edit_revert_subtitle))
                .replace(":revertIntro".toRegex(), context.getString(R.string.description_edit_revert_intro))
                .replace(":revertHistory".toRegex(), context.getString(R.string.description_edit_revert_history, getHistoryUri(qNumber))))
        val revertReason1 = createReason(context.getString(R.string.description_edit_revert_reason1,
            context.getString(R.string.wikidata_description_guide_url)))
        val revertReason2 = createReason(context.getString(R.string.description_edit_revert_reason2))
        binding.helpText.text = SpannableString(TextUtils.expandTemplate(helpStr, revertReason1, revertReason2))
    }

    private fun createReason(htmlString: String) = buildSpannedString {
        inSpans(BulletSpan(DimenUtil.roundedDpToPx(8f))) {
            append(StringUtil.fromHtml(htmlString))
        }
    }

    private fun getHistoryUri(qNumber: String): Uri {
        return Uri.Builder()
                .scheme(WikipediaApp.instance.wikiSite.scheme())
                .authority("m.wikidata.org")
                .appendPath("wiki")
                .appendPath("Special:History")
                .appendPath(qNumber)
                .build()
    }
}
